compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -153,9 +179,29 @@ impl<'thir, 'tcx> Visitor<'thir, 'tcx> for UnsafetyVisitor<'tcx> {
                     self.requires_unsafe(expr.span, CallToUnsafeFunction);
                 }
             }
+            ExprKind::Field { lhs: Expr { ty, .. }, .. } => {
+                // assigning to union field is okay for AccessToUnionField
+                if let ty::Adt(adt_def, _) = ty.kind() {
+                    if adt_def.is_union() {
+                        if self.in_possible_lhs_union_assign {
+                            // FIXME: trigger AssignToDroppingUnionField unsafety if needed
+                        } else {
+                            self.requires_unsafe(expr.span, AccessToUnionField);
+                        }
+                    }
+                }
+            }
+            // don't have any special handling for AssignOp since it causes a read *and* write to lhs
+            ExprKind::Assign { lhs, rhs } => {
+                // assigning to a union is safe, check here so it doesn't get treated as a read later
+                self.in_possible_lhs_union_assign = true;
+                visit::walk_expr(self, lhs);
+                self.in_possible_lhs_union_assign = false;
+                visit::walk_expr(self, rhs);
+                return;
+            }
             _ => {}
         }
-
         visit::walk_expr(self, expr);
```

Comment:
Doesn't this re-walk the LHS and RHS a 2nd time for `ExprKind::Assign`?

Reply:
We return early in the `ExprKind::Assign` branch so this would never be reached in that case.

Reply:
Ah, we do. Could you add a comment about that in said branch?

Reply:
Added in feab574b509668a082e95ea72a1595d8b222c8a0